### PR TITLE
Example of K-fold Cross Validation with `Fabric`

### LINKED ISSUE #16902: K-fold cross validation example in Fabric
### Description & Motivation

Loop customization was removed from Trainer 2.0, and thus solutions like https://github.com/SkafteNicki/pl_cross no longer work. We want to provide examples and answers to users who need to perform k-fold cross validation. 

### Pitch

Build an example how to do [k-fold cross validation](https://sebastianraschka.com/blog/2016/model-evaluation-selection-part3.html#introduction-to-k-fold-cross-validation
) with Fabric. 

Could also be done for Trainer, but is more involved and requires changes to Trainer (probably). An example with Fabric would be a good starting point.

### Additional context

#839
See also Slack and social media for several users who asked for it (not specifically with Fabric, but in general).

cc @borda @carmocca @justusschock @awaelchli

### PR DESCRIPTION
**Pull request by shenoynikhil**
## What does this PR do?

Resolves #16902 

<details>
  <summary><b>Before submitting</b></summary>

- Was this **discussed/agreed** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- Did you make sure to **update the documentation** with your changes? (if necessary)
- Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- Did you list all the **breaking changes** introduced by this pull request?
- Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

<details>
  <summary>Reviewer checklist</summary>

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

</details>

### CODE REVIEW

**Comment by peterchristofferholm** [comment]
Instead of having the k-folds in outer loop, it would be better to loop over the folds in each epoch. Otherwise you will not be able to things like early stopping or pruning doing hyperparameter sweeps. 

```
for epoch in range(max_epochs):
    losses = []
    for k in range(k_folds):
        yhat = model[k](data[k])
        losses[k] = ...
    epoch_loss = mean(losses)
```

**Comment by shenoynikhil** [comment]
>Instead of having the k-folds in outer loop, it would be better to loop over the folds in each epoch. Otherwise you will not be able to things like early stopping or pruning doing hyperparameter sweeps.

That makes sense. I'm looking into it.

**Comment by shenoynikhil** [comment]
@peterchristofferholm I made the changes accordingly. I believe I saw K-Fold implemented in both ways, one where the k-fold loop is the outermost one and one where the epoch loop is. Not sure if there is one correct one, because eventually, you would want to aggregate validation score across the models.

@awaelchli Can you maybe also review how Fabric has been used?

**Comment by justusschock** [comment]
> Instead of having the k-folds in outer loop, it would be better to loop over the folds in each epoch. Otherwise you will not be able to things like early stopping or pruning doing hyperparameter sweeps.

The downside of this is, that with the current implementation you have K-times your model in GPU-memory (which often fits only a single model), or you have to do manual device management

**Comment by shenoynikhil** [comment]
@awaelchli @justusschock @peterchristofferholm  I guess I have both the implementations in the git history. Let me know which one is preferred, and I'll keep that.

File: examples/fabric/kfold_cv/README.md [hunk 1]
```diff
@@ -0,0 +1,45 @@
+## K-Fold Cross Validation
+
+This is an example of performing K-Fold cross validation supported with [Lightning Fabric](https://pytorch-lightning.readthedocs.io/en/stable/fabric/fabric.html). To know more about cross validation, check out [this article](https://sebastianraschka.com/blog/2016/model-evaluation-selection-part3.html#introduction-to-k-fold-cross-validation).
```
Comment (awaelchli) [inline examples/fabric/kfold_cv/README.md:3]:
```suggestion
This is an example of performing K-Fold cross validation supported with [Lightning Fabric](https://pytorch-lightning.readthedocs.io/en/stable/fabric/fabric.html). To learn more about cross validation, check out [this article](https://sebastianraschka.com/blog/2016/model-evaluation-selection-part3.html#introduction-to-k-fold-cross-validation).
```

File: examples/fabric/kfold_cv/README.md [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+
+We use the MNIST dataset to train a simple CNN model. We create the k-fold cross validation splits using the `ModelSelection.KFold` [class](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.KFold.html) in the `scikit-learn` library. Ensure that you have the `scikit-learn` library installed.
```
Comment (awaelchli) [inline examples/fabric/kfold_cv/README.md:5]:
```suggestion
We use the MNIST dataset to train a simple CNN model. We create the k-fold cross validation splits using the `ModelSelection.KFold` [class](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.KFold.html) in the `scikit-learn` library. Ensure that you have the `scikit-learn` library installed:
```
And then below, could you add a block with

```
pip install skcikit-learn
```
So it's easy to copy paste and install :)

File: examples/fabric/kfold_cv/README.md [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+
+### Run
+
+Tip: You can easily inspect the difference between the two files with:
+
+```bash
+sdiff train_torch.py train_fabric.py
+```
```
Comment (awaelchli) [inline examples/fabric/kfold_cv/README.md:13]: I see that you followed the previous examples with fabric and equivalent torch code. The motivation there was to help people who learn about fabric understand the differences between torch and fabric, and to basically show that there are very few of them!

I think when we get to examples like k-fold validation etc. It becomes a bit redundant to have the torch code also. I suggest to drop train_torch.py completely here. This will also be easier to maintain. No strong opinion here, I just think it's cleaner :)

**Review by awaelchli — COMMENTED** [review]
LGTM, minor comments. 
Thanks for jumping on this so quickly. Impressive!

File: examples/fabric/kfold_cv/README.md (hunk 3 above)
Comment (shenoynikhil) [reply examples/fabric/kfold_cv/README.md]: I agree it does not make much sense to add it in this folder. I thought there was some test that calls `train_torch.py` from this folder, so for consistency, I added it. I have removed it and updated README accordingly.

File: examples/fabric/kfold_cv/README.md (hunk 2 above)
Comment (shenoynikhil) [reply examples/fabric/kfold_cv/README.md]: I agree, updated it.

File: examples/fabric/kfold_cv/README.md (hunk 1 above)
Comment (shenoynikhil) [reply examples/fabric/kfold_cv/README.md]: Updated it.

File: examples/fabric/kfold_cv/train_fabric.py [hunk 4]
```diff
@@ -0,0 +1,197 @@
...
+    # to ensure the current process is allowed to save a checkpoint
+    if hparams.save_model:
+        fabric.save(model.state_dict(), "mnist_cnn.pt")
+
+
+if __name__ == "__main__":
+    # Arguments can be passed in through the CLI as normal and will be parsed here
+    # Example:
+    # lightning run model image_classifier.py accelerator=cuda --epochs=3
+    parser = argparse.ArgumentParser(description="Fabric MNIST K-Fold Cross Validation Example")
```
Comment (awaelchli) [inline examples/fabric/kfold_cv/train_fabric.py:178]: how about making the number of folds configurable here?

**Review by awaelchli — APPROVED** [review]

File: examples/fabric/kfold_cv/train_fabric.py (hunk 4 above)
Comment (shenoynikhil) [reply examples/fabric/kfold_cv/train_fabric.py:179]: Added it :).

File: examples/fabric/kfold_cv/train_fabric.py [hunk 5]
```diff
@@ -0,0 +1,199 @@
...
+        x = self.fc1(x)
+        x = F.relu(x)
+        x = self.dropout2(x)
+        x = self.fc2(x)
+        output = F.log_softmax(x, dim=1)
+        return output
+
+
+def train_dataloader(model, data_loader, optimizer, fabric, epoch, hparams, fold):
+    # TRAINING LOOP
```
Comment (jbschiratti) [inline examples/fabric/kfold_cv/train_fabric.py:61]: @shenoynikhil Very naive question (because [I was interested in this feature at some point](https://github.com/Lightning-AI/lightning/issues/839#issuecomment-808489228)): one of the key arguments in favor of lightning is that it removes boilerplate code; then how come you're writing you own training/evaluation loops and not using `pl.Trainer`?

Comment (jbschiratti) [reply examples/fabric/kfold_cv/train_fabric.py:61]: When I read this, I feel like there's really no advantage to using lightning over [accelerate](https://github.com/huggingface/accelerate)! (but there should be, right?) :-)

Comment (shenoynikhil) [reply examples/fabric/kfold_cv/train_fabric.py:61]: So there are two ways to include lightning in your code. First, is using the PL framework end-to-end (Training, LightningModule, etc), this removes most of the boilerplate code as you said and allows you to do advanced stuff like distributed training, etc. Second, is using lightning [fabric](https://pytorch-lightning.readthedocs.io/en/stable/fabric/fabric.html) which allows you to modify your existing pytorch code to enable it many of the advantages provided by PL (read the documentation for more info on how this is different from the first part).

Now coming to KFold cross validation, implementing this the first way is a bit challenging since it requires changes to the `pl.Trainer`. So an easier way to still power KFold cross validation with pl would be to use fabric. @awaelchli might be able to add more on this :). Hope I have still answered some part of your question.

Comment (jbschiratti) [reply examples/fabric/kfold_cv/train_fabric.py:61]: Understood, thanks! I believe that the real added value would be to make `pl.Trainer` work with CV (or nested CV). As you noted, that's the complicated part...

**Review by Borda — APPROVED** [review]
